File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.cpp
class Solution {
public:
    const int mod = 1e9+7;
    int numFactoredBinaryTrees(vector<int>& arr) {
        sort(begin(arr),end(arr));
        int n = arr.size();
        vector<int64_t>dp(n,1);
        unordered_map<int,int>us;
        for(int i=0;i<n;++i){
            us[arr[i]]=i;
        }
        int64_t ans =0;
        for(int i=1;i<n;++i){
            int64_t temp = 0;
            for(int j=0;j<i;++j){
                if((arr[i]%arr[j]==0 )and us.find(arr[i]/arr[j])!=us.end()){
                    temp += ((dp[us[arr[i]/arr[j]]]*dp[j])%mod);
                }
            }
            dp[i]+=(temp)%mod;
        }
        return accumulate(begin(dp),end(dp),0ll)%mod;
    }
};

File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        arr.sort()
        mod = 1e9 + 7
        n = len(arr)
        dp = [ 1 for i in range(n)]
        dic = {}
        for i in range(n):
            dic[arr[i]] = i
        for i in range(1,n):
            temp = 0 
            for j in range(0,i):
                if arr[i]%arr[j]==0 :
                    x = arr[i]//arr[j]
                    if x in dic:
                        temp += int ((dp[j]*dp[dic[x]])%mod)
            temp %= mod 
            dp[i] += temp
        ans = 0 
        for i in dp:
            ans += i
        return int (ans%mod )
        